test/unit/html/patient_round_trip_test.rb

```diff
@@ -0,0 +1,167 @@
+require_relative '../../test_helper'
+require 'cqm/models'
+require 'byebug'
+
+module HTML
+  class PatientRoundTripTest < MiniTest::Test
+    # include QRDA::Cat1
+
+    def setup
+      create_test_measures_collection
+      address = CQM::Address.new(
+        use: 'HP',
+        street: ['202 Burlington Rd.'],
+        city: 'Bedford',
+        state: 'MA',
+        zip: '01730',
+        country: 'US'
+      )
+      telecom = CQM::Telecom.new(
+        use: 'HP',
+        value: '555-555-2003'
+      )
+      @options = { start_time: Date.new(2012, 1, 1), end_time: Date.new(2012, 12, 31), patient_addresses: [address], patient_telecoms: [telecom] }
+      # TODO: use address etc??
+
+      generate_shell_patient('html') # builds @cqm_patient and @qdm_patient
+    end
+
+    def test_demographics
+      @cqm_patient.qdmPatient = @qdm_patient
+      html = QdmPatient.new(@cqm_patient, true).render
+      %w[gender race ethnicity birthdate payer].each do |cat|
+        assert html.include?(@qdm_patient.get_data_elements('patient_characteristic', cat).first.dataElementCodes.first[:code])
+      end
+    end
+
+    def test_all_html_attributes
+      qdm_types = YAML.safe_load(File.read(File.expand_path('../../../config/qdm_types.yml', __dir__)), [], [], true)
+      qdm_types.each do |qt|
+        dt = QDM::PatientGeneration.generate_loaded_datatype("QDM::#{qt}")
+        # check custom generated datatype for negationRationale field
+        check_loaded_patient(QDM::PatientGeneration.generate_loaded_datatype("QDM::#{qt}", true), "negationRationale", qt) if dt.typed_attributes.keys.include?("negationRationale")
+        # iterate through all relevant attributes for each data type, except negationRationale (checked above)
+        dt.typed_attributes.keys.each do |field|
+          next if %w[_id dataElementCodes description codeListId id _type type qdmTitle hqmfOid qrdaOid qdmCategory qdmVersion qdmStatus negationRationale targetOutcome].include? field
+          # TODO: test targetOutcome (in Care Goal, not currently exported)
+          check_loaded_patient(dt, field, qt)
+        end
+      end
+    end
+
+    def check_loaded_patient(dt, field, qt)
+      # create new qdm patient clone for dt/attribute combo
+      qdm_patient = qdm_patient_for_attribute(dt, field, @qdm_patient)
+      @cqm_patient.qdmPatient = qdm_patient
+
+      html = QdmPatient.new(@cqm_patient, true).render
+      assert html.include?(qt), "html should include QDM type #{qt}"
+      assert dt.respond_to?(field), "datatype generation discrepancy, should contain field #{field}"
+      check_for_type(html, dt, field)
+    end
+
+    def check_for_type(html, dt, field)
+      attr = dt.send(field)
+      if attr.respond_to?(:strftime)
+        # timey object
+        formatted_date = attr.localtime.strftime('%FT%T')
+        assert html.include?(formatted_date), "html should include date/time value #{formatted_date}"
```

Comment:
Could these assertions also include the datatype and field

for example:

"html should include date/time value #{formatted_date} for #{field} in #{dt._type}"

Might make debugging easier